drivers/usb/device/usb_dc_it82xx2.c

```diff
@@ -347,158 +305,176 @@ static int it82xx2_usb_fifo_ctrl(uint8_t ep)
 	return ret;
 }
 
-/*
- * Functions it82xx2_epn0n1_ext_ctrl_cfg() and epn0n1_ext_ctrl_cfg_seq_inv()
- * provide the entrance of configuring the EPN0N1 Extended Ctrl Registers.
- *
- * The variable set_clr determines if we set/clear the corresponding bit.
- */
-static void it82xx2_epn0n1_ext_ctrl_cfg(uint8_t reg_idx, uint8_t bit_mask,
-			bool set_clr)
-{
-	struct usb_it82xx2_regs *const usb_regs =
-		(struct usb_it82xx2_regs *)it82xx2_get_usb_regs();
-	volatile uint8_t *epn0n1_ext_ctrl =
-		usb_regs->fifo_regs[EP_EXT_REGS_9X].ext_4_15.epn0n1_ext_ctrl;
-
-	(set_clr) ? (epn0n1_ext_ctrl[reg_idx] |= bit_mask) :
-		(epn0n1_ext_ctrl[reg_idx] &= ~(bit_mask));
-}
-
-static void it82xx2_epn0n1_ext_ctrl_cfg_seq_inv(uint8_t reg_idx,
-			uint8_t bit_mask, bool set_clr)
+static int it82xx2_usb_extend_ep_ctrl(uint8_t ep, enum it82xx2_ep_ctrl ctrl, bool enable)
 {
 	struct usb_it82xx2_regs *const usb_regs =
 		(struct usb_it82xx2_regs *)it82xx2_get_usb_regs();
-	volatile uint8_t *epn0n1_ext_ctrl =
+	union epn0n1_extend_ctrl_reg *epn0n1_ext_ctrl =
 		usb_regs->fifo_regs[EP_EXT_REGS_9X].ext_4_15.epn0n1_ext_ctrl;
+	struct epn_ext_ctrl_regs *ext_ctrl =
+		usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
+	uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	uint8_t idx;
 
-	bool check = (set_clr) ? (epn0n1_ext_ctrl[reg_idx] & EPN1_OUTDATA_SEQ)
-			       : (epn0n1_ext_ctrl[reg_idx] & EPN0_OUTDATA_SEQ);
-
-	(check) ? (epn0n1_ext_ctrl[reg_idx] &= ~(bit_mask)) :
-		(epn0n1_ext_ctrl[reg_idx] |= bit_mask);
-}
-
-/* Return the status of STALL bit in the EPN0N1 Extend Control Registers */
-static bool it82xx2_epn01n1_check_stall(uint8_t reg_idx, uint8_t bit_mask)
-{
-	struct usb_it82xx2_regs *const usb_regs =
-		(struct usb_it82xx2_regs *)it82xx2_get_usb_regs();
-	volatile uint8_t *epn0n1_ext_ctrl =
-		usb_regs->fifo_regs[EP_EXT_REGS_9X].ext_4_15.epn0n1_ext_ctrl;
+	if (!IT8XXX2_IS_EXTEND_ENDPOINT(ep_idx)) {
+		return -EINVAL;
+	}
 
-	return !!(epn0n1_ext_ctrl[reg_idx] & bit_mask);
-}
+	if ((ctrl == EP_IN_DIRECTION_SET) || (ctrl == EP_ENABLE) || (ctrl == EP_READY_ENABLE)) {
+		idx = ((ep_idx - EP4) % 3) + 1;
+	} else {
+		idx = (ep_idx - EP4) / 2;
+	}
 
-/* Configuring the EPN Extended Ctrl Registers. */
-static void it82xx2_epn_ext_ctrl_cfg1(uint8_t reg_idx, uint8_t bit_mask,
-			bool set_clr)
-{
-	struct usb_it82xx2_regs *const usb_regs =
-		(struct usb_it82xx2_regs *)it82xx2_get_usb_regs();
-	struct epn_ext_ctrl_regs *epn_ext_ctrl =
-		usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
+	switch (ctrl) {
+	case EP_STALL_SEND:
+		if (ep_idx % 2) {
+			epn0n1_ext_ctrl[idx].fields.epn1_send_stall_bit = enable;
```

Comment:
Please refactor it to get `fields` based on ep_idx and maybe other arguments.

Reply:
Completed. Thanks.